### DISCUSSION TITLE
Persisting Database Files

### DISCUSSION THREAD

**Discussion by psychemedia** [mmh352/ou-container-builder#36] [Ideas]
In the TM351, we currently run the database files inside the container they run locally.

This has the advantage that if a student mangles the db, they can just destroy the container and fire up a new on.

In notebooks where students do write to the database, we tend to have cleanup and init scripts that can try to get small db tables into a the state we need them to be in for a particular's notebook's activities.
 
For assessment, students do need to load files into the db themselves, but part of the point of the module is writing reproducible scripts, so if they mangle the db and need to create a new one, they should have created reusable scripts to load their assessment data back in!;-)

Generally, though, it would make sense to persist database files in `$HOME/$CODE-$PRESENTATION/.postgres/` etc, although this would have a slight impact on storage requirements (which could add up on large population modules).

The official postgres container, and the official mongo container, both let you put scripts into a `/docker-entrypoint-initdb.d/` dir that are handled on first run if a db is not found where the db files are mounted [[code](https://github.com/docker-library/postgres/blob/14f13e4b399ed1848fa24c2c1f5bd40c25732bdd/docker-entrypoint.sh#L292)]. This lets you set up and seed the database in a mounted volume.

You maybe already have some related machinery to handle the `content` (I've not checked the code for that yet...) that does a first run thing that could be copied for first run build related activity? Maybe a `firstrunscript` block would be handy? Or if `scripts` can support multiple script blocks, allow a script block to have a `firstrun` boolean (default False) that lets the script run if a `runalready` flag can't be found?

**Comment by psychemedia** [comment]
I am assuming that under kubernetes, a Dockerfile can't be used to create "arbitrary" persistent volumes when containers are launched from the image?

**Reply by mmh352** [reply]
Not, as far as I am aware on a per-container basis.

**Comment by mmh352** [comment]
So we need some way to run custom scripts on container start (#41).

Telling postgres to store its database files somewhere in /home/ou/$CODE-$PRESENTATION shouldn't be a problem. An alternative would be to have custom scripts to be run on shutdown as well and then you could dump the current DB state to the persistent storage on shutdown and reload it on startup. Thoughts?

**Reply by psychemedia** [reply]
My feeling for some time, though I've never had the pressing need to try it out, was to:

- use `$HOME/.postgres` (so `/home/ou/$CODE-$PRESENTATION/.postgres`) as the db files location;
- have an equivalent to the initdb thing so that a container on first run would check something existed, and if it didn't, then run an initialisation step.

In part I've been waiting for other possible applications of the same idea before trying it, because I get the feeling that a general purpose init/first run thing would be handy. 

Eg in TM351, seeding postgres and mongo.

Copying content over from a stash inside the container is another first run thing.

So this boils down to:

- a stash of files somewhere (content, db setup and seeding files)
- possibly a target for them; possibly a script to run over them;
- some pattern or state to check for;
- an action (eg copy the files, and/or run the script, then set the firstrun flag as having occurred);

If the thing to check for might most sensibly be something standardised that is set and tested automatically, eg the existence of `/home/ou/$CODE-$PRESENTATION/.postgres/.firstrun` because that then gives us a possible reset method eg in the form of telling students to delete `/home/ou/$CODE-$PRESENTATION/.postgres/.firstrun` ?

**Reply by mmh352** [reply]
The files part of this is already there. It currently doesn't have an option to copy something only once, but that could be added.

I'll let you know as soon as the documentation is updated, so that you can see what is already supported.

**Reply by psychemedia** [reply]
"It currently doesn't have an option to copy something only once, but that could be added." How does that work then? With the TM129 container I thought you copied stashed files over to the persisted volume on first run? Or,, ah, do you try to do it every time but have a no-overwrite? (Hmmm... if so, what if a student deletes a file; do you then copy it back next time they start a session?)

**Reply by mmh352** [reply]
Yes. Currently the overwrite options are "always", "never", and "if-unchanged": file:///home/mhall/Projects/ou-container-builder/docs/_build/html/configuration.html#content-to-distribute

If the student removes something, then it will be re-created on the next run.

**Reply by psychemedia** [reply]
Just had this suggested to me as a *de facto* first run thing on the JupyterHub side — https://zero-to-jupyterhub.readthedocs.io/en/latest/jupyterhub/customizing/user-environment.html#about-user-storage-and-adding-files-to-it —which could make sense if:

1. each module-presentation had its own jupyterhub;
2. we didn't have to consider users running their own containers.

**Reply by mmh352** [reply]
It is the second point which is the bigger show-stopper, because it rules out running anything in the lifecycleHooks. The solution we currently have is very close to what nbgitpuller provides, with the advantage that all materials are in the container, thus there is no need to keep two repositories in sync. The one big difference is that if the master copy and the student copy have changes, then the ou-container-content doesn't try to merge those, it just leaves the student copy as-is. Obviously that could be changed, it wouldn't be that hard to implement.

**Reply by psychemedia** [reply]
The key take away for me is that there is a JupyterHub mechanism for content copy.
